builder/llpcTargetInfo.h

```diff
@@ -0,0 +1,164 @@
+            uint32_t  miscSpiSgprsNum                   :  1;
+            uint32_t  shader8b16bLocalWriteCorruption   :  1;
+            uint32_t  shaderCoalesceStore               :  1;
+            uint32_t  shaderEstimateRegisterUsage       :  1;
+            uint32_t  shaderReadlaneSmrd                :  1;
+            uint32_t  shaderSmemBufferAddrClamp         :  1;
+            uint32_t  shaderSpiBarrierMgmt              :  1;
+            //
+            //
+            //
+            //
+            //
+
+            uint32_t  shaderSpiCsRegAllocFragmentation  :  1;
+            uint32_t  shaderVcczScalarReadBranchFailure :  1;
+            uint32_t  shaderZExport                     :  1;
+            // Pre-GFX9 hardware doesn't support min/max denorm flush, we insert extra fmul with 1.0 to flush the denorm value
+            uint32_t  shaderMinMaxFlushDenorm           :  1;
+            uint32_t  reserved                          : 19;
+        };
+        uint32_t  u32All;
+    } gfx6;
+
+    union
+    {
+        struct
+        {
+            uint32_t  fixCacheLineStraddling       :  1;
+            uint32_t  fixLsVgprInput               :  1;
+            uint32_t  shaderImageGatherInstFix     :  1;
+            uint32_t  treat1dImagesAs2d            :  1;
+            uint32_t  reserved                     : 28;
+        };
+        uint32_t  u32All;
+    } gfx9;
+
+#if LLPC_BUILD_GFX10
+    union
+    {
+        struct
+        {
+            uint32_t  disableI32ModToI16Mod        :  1;
+
+            uint32_t  waTessFactorBufferSizeLimitGeUtcl1Underflow : 1;
+            uint32_t  waTessIncorrectRelativeIndex : 1;
+            uint32_t  waShaderInstPrefetch123 : 1;
+            uint32_t  waShaderInstPrefetch0 : 1;
+            uint32_t  nggTessDegeneratePrims : 1;
+            uint32_t  waDidtThrottleVmem : 1;
+            uint32_t  waLdsVmemNotWaitingVmVsrc : 1;
+            uint32_t  waNsaCannotBeLastInClause : 1;
+            uint32_t  waNsaAndClauseCanHang : 1;
+            uint32_t  waNsaCannotFollowWritelane : 1;
+            uint32_t  waThrottleInMultiDwordNsa : 1;
+            uint32_t  waSmemFollowedByVopc : 1;
+            uint32_t  waNggCullingNoEmptySubgroups : 1;
+            uint32_t  waShaderInstPrefetchFwd64 : 1;
+            uint32_t  waWarFpAtomicDenormHazard : 1;
+            uint32_t  waNggDisabled : 1;
+            uint32_t  reserved : 15;
+        };
+        uint32_t u32All;
+    } gfx10;
+#endif
+};
+
+// =====================================================================================================================
+// TargetInfo class, representing features and workarounds for the particular selected target
+class TargetInfo
+{
+public:
+    // Set TargetInfo. Returns false if the GPU name is not found or not supported.
+    bool SetTargetInfo(StringRef gpuName);
+
+    // Accessors.
+    GfxIpVersion GetGfxIpVersion() const { return m_gfxIp; }
```

Comment:
I think we should do that in the same way that the LLVM AMDGPU backend does it: gfx version is not mentioned at all in the code, and instead there is a set of features to test. For example, to see if shaders are merged, you would check the "merged shaders" feature instead of checking gfxip >= 9.

Reply:
This could be a good idea, introducing a set of feature checker "SupportsXXX()".